toolchain/check/generic.cpp

```diff
@@ -61,60 +63,48 @@ static auto AddGenericConstantInstToEvalBlock(
 }
 
 namespace {
-// A map from an instruction ID representing a canonical symbolic constant to an
-// instruction within an eval block of the generic that computes the specific
-// value for that constant.
-//
-// We arbitrarily use a small size of 256 bytes for the map.
-// TODO: Determine a better number based on measurements.
-using ConstantsInGenericMap = Map<SemIR::InstId, SemIR::InstId, 256>;
-
 // Substitution callbacks to rebuild a generic constant in the eval block for a
 // generic region.
 class RebuildGenericConstantInEvalBlockCallbacks : public SubstInstCallbacks {
  public:
   // `context` must not be null.
-  RebuildGenericConstantInEvalBlockCallbacks(
-      Context* context, SemIR::GenericId generic_id,
-      SemIR::GenericInstIndex::Region region, SemIR::LocId loc_id,
-      ConstantsInGenericMap& constants_in_generic, bool inside_redeclaration)
-      : context_(context),
-        generic_id_(generic_id),
-        region_(region),
+  RebuildGenericConstantInEvalBlockCallbacks(Context* context,
+                                             SemIR::LocId loc_id)
+      : SubstInstCallbacks(context),
         loc_id_(loc_id),
-        constants_in_generic_(constants_in_generic),
-        inside_redeclaration_(inside_redeclaration) {}
-
-  auto context() const -> Context& { return *context_; }
+        constants_in_generic_(
+            context->generic_region_stack().PeekConstantsInGenericMap()) {}
+
+  auto RebuildType(SemIR::TypeInstId type_inst_id) const
+      -> SemIR::TypeId override {
+    // When building instructions in the eval block, form attached types.
+    return context().types().GetTypeIdForTypeConstantId(
+        context().constant_values().GetAttached(type_inst_id));
+  }
 
   // Check for instructions for which we already have a mapping into the eval
   // block, and substitute them for the instructions in the eval block.
   auto Subst(SemIR::InstId& inst_id) const -> bool override {
-    auto const_id = context_->constant_values().Get(inst_id);
+    auto const_id = context().constant_values().Get(inst_id);
     if (!const_id.has_value()) {
       // An unloaded import ref should never contain anything we need to
       // substitute into. Don't trigger loading it here.
       CARBON_CHECK(
-          context_->insts().Is<SemIR::ImportRefUnloaded>(inst_id),
+          context().insts().Is<SemIR::ImportRefUnloaded>(inst_id),
           "Substituting into instruction with invalid constant ID: {0}",
-          context_->insts().Get(inst_id));
+          context().insts().Get(inst_id));
       return true;
     }
-    if (!context_->constant_values().DependsOnGenericParameter(const_id)) {
+    if (!context().constant_values().DependsOnGenericParameter(const_id)) {
       // This instruction doesn't have a symbolic constant value, so can't
       // contain any bindings that need to be substituted.
       return true;
     }
 
     // If this instruction is in the map, return the known result.
```

Comment:
```suggestion
    // If this constant value has a defining instruction in the eval block, replace the
    // instruction in the body of the generic with the one from the eval block.
```
Is this right then? It would be helpful I think to explain the relationship here in terms of the instructions instead of just the fact we use a map.